File: langs/python/src/vmx/services/message_hub.py
"""Message hub — Protocol and concrete Subject-backed implementation."""

from __future__ import annotations

import logging
from collections import deque
from collections.abc import Iterator
from contextlib import AbstractContextManager, contextmanager
from threading import Condition, RLock, get_ident
from typing import Generic, Protocol, TypeVar, runtime_checkable

import reactivex as rx
from reactivex.abc import DisposableBase, ObserverBase
from reactivex.subject import Subject

from vmx.messages.protocols import Message

_logger = logging.getLogger(__name__)


_wait_graph_gate = RLock()
_thread_waits_for: dict[int, int] = {}


def _register_message_hub_wait(waiter: int, owner: int) -> bool:
    """Register one wait edge; return true if it would close a real cycle."""
    with _wait_graph_gate:
        _thread_waits_for[waiter] = owner
        cursor = owner
        visited: set[int] = set()
        while cursor not in visited:
            if cursor == waiter:
                _thread_waits_for.pop(waiter, None)
                return True
            visited.add(cursor)
            next_owner = _thread_waits_for.get(cursor)
            if next_owner is None:
                return False
            cursor = next_owner
        return False


def _unregister_message_hub_wait(waiter: int) -> None:
    with _wait_graph_gate:
        _thread_waits_for.pop(waiter, None)


TMessage = TypeVar("TMessage", bound=Message, contravariant=True)


@runtime_checkable
class MessageHubProto(Protocol[TMessage]):
    """Hot pub/sub stream for IMessage events.

    Structural Protocol — any class that provides ``.messages`` and ``.send()``
    satisfies this interface without explicit inheritance.
    """

    @property
    def messages(self) -> rx.Observable[Message]:
        """Observable stream of all messages published to this hub."""
        ...

    def send(self, message: TMessage) -> None:
        """Publish *message* to all current subscribers."""
        ...


@runtime_checkable
class TransactionalMessageHubProto(MessageHubProto[TMessage], Protocol[TMessage]):
    """Additive capability for hubs that support message transactions."""

    def batch(self) -> AbstractContextManager[None]:
        """Return a synchronous, nestable message transaction scope."""
        ...


_THubMessage = TypeVar("_THubMessage", bound=Message)


class MessageHub(Generic[_THubMessage]):
    """Default Subject-backed hub.  Hot stream — no replay buffer.

    Subscriber-handler exceptions are isolated per-subscription so a raising
    handler does not terminate the stream for other subscribers (HUB-007).
    """

    def __init__(self) -> None:
        self._subject: Subject[Message] = Subject()
        self._gate = Condition(RLock())
        self._pending: deque[Message] = deque()
        self._disposed: bool = False
        self._subject_termination_claimed = False
        self._subject_terminated = False
        self._drainer_thread: int | None = None
        self._batch_owner: int | None = None
        self._batch_depth = 0
        self._borrowed_batch_depth = 0

    def _wait_for_owner_locked(self, caller: int, owner: int) -> bool:
        """Wait once for *owner*, or report that waiting would close a cycle."""
        if _register_message_hub_wait(caller, owner):
            return True
        try:
            self._gate.wait()
        finally:
            _unregister_message_hub_wait(caller)
        return False

    @property
    def messages(self) -> rx.Observable[Message]:
        """Per HUB-007: wrap each subscription so exceptions are swallowed."""
        return rx.create(self._subscribe_safely)

    def _subscribe_safely(
        self,
        observer: ObserverBase[Message],
        scheduler: object = None,
    ) -> DisposableBase:
        def on_next(value: Message) -> None:
            try:
                observer.on_next(value)
            except Exception:
                # Isolate the failing subscriber (HUB-007): the exception is
                # swallowed so it cannot terminate the stream for other
                # subscribers (spec/03-messages.md §Subscriber resilience). It
                # is logged (not silently discarded) so the failure is still
                # observable to operators via the logging framework.
                _logger.exception(
                    "MessageHub subscriber raised while handling %r; isolating per HUB-007",
                    type(value).__name__,
                )

        return self._subject.subscribe(
            on_next=on_next,
            on_error=observer.on_error,
            on_completed=observer.on_completed,
        )

    def send(self, message: _THubMessage) -> None:
        """Publish *message* synchronously to all current subscribers."""
        caller = get_ident()
        should_drain = False
        with self._gate:
            while not self._disposed:
                owner = (
                    self._batch_owner
                    if self._batch_owner not in (None, caller)
                    else self._drainer_thread
                    if self._drainer_thread not in (None, caller)
                    else None
                )
                if owner is not None:
                    if self._wait_for_owner_locked(caller, owner):
                        break
                    continue
                if self._borrowed_batch_depth > 0:
                    self._gate.wait()
                    continue
                break
            if self._disposed:
                return
            self._pending.append(message)
            if self._batch_owner is not None or self._borrowed_batch_depth > 0:
                return
            if self._drainer_thread is None:
                self._drainer_thread = caller
                should_drain = True
            elif self._drainer_thread == caller:
                return

        if should_drain:
            self._drain()

    @contextmanager
    def batch(self) -> Iterator[None]:
        """Defer messages until the outermost transaction scope exits."""
        caller = get_ident()
        entered = False
        borrowed = False
        with self._gate:
            while not self._disposed:
                owner = (
                    self._batch_owner
                    if self._batch_owner not in (None, caller)
                    else self._drainer_thread
                    if self._drainer_thread not in (None, caller)
                    else None
                )
                if owner is not None:
                    if self._wait_for_owner_locked(caller, owner):
                        self._borrowed_batch_depth += 1
                        borrowed = True
                        break
                    continue
                if self._borrowed_batch_depth > 0:
                    self._gate.wait()
                    continue
                break
            if not self._disposed:
                if not borrowed:
                    self._batch_owner = caller
                    self._batch_depth += 1
                entered = True
        body_error: BaseException | None = None
        try:
            yield
        except BaseException as error:
            body_error = error
        should_drain = False
        should_terminate_subject = False
        if entered:
            with self._gate:
                if borrowed:
                    self._borrowed_batch_depth -= 1
                    outermost = self._borrowed_batch_depth == 0
                else:
                    self._batch_depth -= 1
                    outermost = self._batch_depth == 0
                    if outermost:
                        self._batch_owner = None
                if outermost:
                    if (
                        not self._disposed
                        and self._pending
                        and self._drainer_thread is None
                        and self._batch_owner is None
                        and self._borrowed_batch_depth == 0
                    ):
                        self._drainer_thread = caller
                        should_drain = True
                    elif (
                        self._disposed
                        and self._drainer_thread is None
                        and self._batch_owner is None
                        and self._borrowed_batch_depth == 0
                        and not self._subject_termination_claimed
                    ):
                        self._subject_termination_claimed = True
                        should_terminate_subject = True
                    self._gate.notify_all()
        drain_error: BaseException | None = None
        if should_drain:
            try:
                self._drain()
            except BaseException as error:
                drain_error = error
        if should_terminate_subject:
            self._terminate_subject()
        if body_error is not None:
            raise body_error
        if drain_error is not None:
            raise drain_error

    def _drain(self) -> None:
        delivered = 0
        message_types: set[str] = set()
        while True:
            stop_draining = False
            should_terminate_subject = False
            with self._gate:
                while self._borrowed_batch_depth > 0 and not self._disposed:
                    self._gate.wait()
                if self._disposed or not self._pending:
                    should_terminate_subject = self._release_drainer_locked()
                    stop_draining = True
                    message = None
                else:
                    message = self._pending.popleft()
            if should_terminate_subject:
                self._terminate_subject()
            if stop_draining:
                return
            if __debug__:
                assert message is not None
                message_types.add(type(message).__name__)
            try:
                assert message is not None
                self._subject.on_next(message)
            except BaseException:
                try:
                    self._abandon_pending()
                except BaseException:
                    pass  # Preserve the original delivery failure.
                raise
            if __debug__:
                delivered += 1
                with self._gate:
                    has_pending = bool(self._pending)
                if delivered >= 10_000 and has_pending:
                    with self._gate:
                        message_types.update(type(item).__name__ for item in self._pending)
                    names = ", ".join(sorted(message_types))
                    cycle_error = RuntimeError(
                        "MessageHub drain exceeded 10000 messages; "
                        f"possible publish cycle involving: {names}"
                    )
                    self._abandon_pending()
                    raise cycle_error

    def _abandon_pending(self) -> None:
        should_terminate_subject = False
        with self._gate:
            self._pending.clear()
            should_terminate_subject = self._release_drainer_locked()
        if should_terminate_subject:
            self._terminate_subject()

    def _release_drainer_locked(self) -> bool:
        """Release the drainer and claim deferred subject termination."""
        self._drainer_thread = None
        should_terminate_subject = self._disposed and not self._subject_termination_claimed
        if should_terminate_subject:
            self._subject_termination_claimed = True
        self._gate.notify_all()
        return should_terminate_subject

    def _terminate_subject(self) -> None:
        try:
            self._subject.on_completed()
        finally:
            try:
                self._subject.dispose()
            finally:
                with self._gate:
                    self._subject_terminated = True
                    self._gate.notify_all()

    def dispose(self) -> None:
        """Complete and dispose the underlying subject."""
        caller = get_ident()
        should_terminate_subject = False
        with self._gate:
            while not self._disposed:
                owner = (
                    self._batch_owner
                    if self._batch_owner not in (None, caller)
                    else self._drainer_thread
                    if self._drainer_thread not in (None, caller)
                    else None
                )
                if owner is not None:
                    if self._wait_for_owner_locked(caller, owner):
                        self._disposed = True
                        self._pending.clear()
                        self._gate.notify_all()
                        return
                    continue
                if self._borrowed_batch_depth > 0:
                    self._gate.wait()
                    continue
                break
            if self._disposed:
                return
            self._disposed = True
            self._pending.clear()
            if self._drainer_thread is None:
                self._subject_termination_claimed = True
                should_terminate_subject = True
            self._gate.notify_all()
        if should_terminate_subject:
            self._terminate_subject()

File: langs/csharp/src/VMx/Services/MessageHub.cs
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Runtime.ExceptionServices;
using VMx.Internal;
using VMx.Messages;

namespace VMx.Services;

/// <summary>
/// Default Subject-backed hub. Hot stream — late subscribers do not see
/// prior messages. Subscriber-handler exceptions are swallowed (per HUB-007).
/// </summary>
public sealed class MessageHub : ITransactionalMessageHub, IDisposable
{
#if DEBUG
    private const int DevelopmentDrainLimit = 10_000;
#endif
    [ThreadStatic]
    private static int s_deliveryDepth;
    private static readonly object s_waitGraphGate = new();
    private static readonly Dictionary<int, int> s_waitGraph = [];
    private readonly object _gate = new();
    private readonly Queue<IMessage> _pending = new();
    private readonly Subject<IMessage> _subject = new();
    private bool _disposeRequested;
    private bool _disposed;
    private bool _subjectTerminationClaimed;
    private int _drainerThreadId;
    private int _batchOwnerThreadId;
    private int _batchDepth;
    private int _borrowedBatchDepth;

    /// <inheritdoc/>
    public IObservable<IMessage> Messages =>
        Observable.Create<IMessage>(observer =>
            _subject.Subscribe(
                onNext: message =>
                {
                    try { observer.OnNext(message); }
                    catch { /* isolate the failing subscriber per HUB-007 */ }
                },
                onError: observer.OnError,
                onCompleted: observer.OnCompleted));

    /// <inheritdoc/>
    public void Send<TMessage>(TMessage message) where TMessage : IMessage
    {
        var caller = Environment.CurrentManagedThreadId;
        var shouldDrain = false;
        lock (_gate)
        {
            while (!_disposed && !_disposeRequested)
            {
                var owner = ConflictingOwnerLocked(caller);
                if (owner != 0)
                {
                    // Only an edge that closes a real cross-hub wait cycle may
                    // defer this send into the target's existing FIFO drain.
                    if (WaitForOwnerLocked(caller, owner)) break;
                    continue;
                }
                if (_borrowedBatchDepth > 0)
                {
                    if (s_deliveryDepth > 0) break;
                    Monitor.Wait(_gate);
                    continue;
                }
                break;
            }
            if (_disposed || _disposeRequested) return;

            _pending.Enqueue(message);
            if (_batchOwnerThreadId != 0 || _borrowedBatchDepth > 0) return;
            if (_drainerThreadId == 0)
            {
                _drainerThreadId = caller;
                shouldDrain = true;
            }
            else if (_drainerThreadId == caller)
            {
                return;
            }
        }

        if (shouldDrain) DrainQueue();
    }

    /// <inheritdoc/>
    public void Batch(Action transaction)
    {
        ThrowHelper.ThrowIfNull(transaction, nameof(transaction));
        var caller = Environment.CurrentManagedThreadId;
        var entered = false;
        var borrowed = false;
        lock (_gate)
        {
            while (!_disposed && !_disposeRequested)
            {
                var owner = ConflictingOwnerLocked(caller);
                if (owner != 0)
                {
                    if (WaitForOwnerLocked(caller, owner))
                    {
                        // A cyclic batch borrows the target: its body may enqueue,
                        // but the target owner cannot resume draining until this
                        // borrowed scope exits.
                        _borrowedBatchDepth++;
                        borrowed = true;
                        entered = true;
                        break;
                    }
                    continue;
                }
                if (_borrowedBatchDepth > 0 && _batchOwnerThreadId != caller)
                {
                    Monitor.Wait(_gate);
                    continue;
                }
                break;
            }
            if (!_disposed && !_disposeRequested && !borrowed)
            {
                _batchOwnerThreadId = caller;
                _batchDepth++;
                entered = true;
            }
        }

        ExceptionDispatchInfo? callbackError = null;
        try { transaction(); }
        catch (Exception error) { callbackError = ExceptionDispatchInfo.Capture(error); }

        var shouldDrain = false;
        var shouldTerminateSubject = false;
        if (entered)
        {
            lock (_gate)
            {
                bool outermost;
                if (borrowed)
                {
                    _borrowedBatchDepth--;
                    outermost = _borrowedBatchDepth == 0;
                }
                else
                {
                    _batchDepth--;
                    outermost = _batchDepth == 0;
                    if (outermost) _batchOwnerThreadId = 0;
                }

                if (outermost && _disposeRequested)
                    shouldTerminateSubject = FinishDisposeLocked();

                if (outermost && !_disposed && _pending.Count > 0 &&
                    _batchOwnerThreadId == 0 && _borrowedBatchDepth == 0 &&
                    _drainerThreadId == 0)
                {
                    _drainerThreadId = caller;
                    shouldDrain = true;
                }
                if (outermost)
                    Monitor.PulseAll(_gate);
            }
        }

        if (shouldTerminateSubject) TerminateSubject();
        ExceptionDispatchInfo? drainError = null;
        if (shouldDrain)
        {
            try { DrainQueue(); }
            catch (Exception error) { drainError = ExceptionDispatchInfo.Capture(error); }
        }
        callbackError?.Throw();
        drainError?.Throw();
    }

    private void DrainQueue()
    {
#if DEBUG
        var delivered = 0;
        var messageTypes = new HashSet<string>(StringComparer.Ordinal);
#endif
        while (true)
        {
            IMessage? message = null;
            var stopDraining = false;
            var shouldTerminateSubject = false;
            lock (_gate)
            {
                while (_borrowedBatchDepth > 0 && !_disposed)
                    Monitor.Wait(_gate);
                if (_disposeRequested)
                    _ = FinishDisposeLocked();
                if (_disposed || _pending.Count == 0)
                {
                    shouldTerminateSubject = ReleaseDrainerLocked();
                    stopDraining = true;
                }
                else
                {
                    message = _pending.Dequeue();
                }
            }
            if (shouldTerminateSubject) TerminateSubject();
            if (stopDraining) return;
#if DEBUG
            messageTypes.Add(message!.GetType().Name);
#endif
            try
            {
                s_deliveryDepth++;
                _subject.OnNext(message!);
            }
            catch
            {
                try { AbandonPending(); }
                catch { /* preserve the original delivery failure */ }
                throw;
            }
            finally
            {
                s_deliveryDepth--;
            }
#if DEBUG
            delivered++;
            lock (_gate)
            {
                if (delivered < DevelopmentDrainLimit || _pending.Count == 0) continue;
                foreach (var pending in _pending)
                    messageTypes.Add(pending.GetType().Name);
            }
            var cycleError = new InvalidOperationException(
                $"MessageHub drain exceeded {DevelopmentDrainLimit} messages; " +
                $"possible publish cycle involving: " +
                $"{string.Join(", ", messageTypes.OrderBy(name => name, StringComparer.Ordinal))}");
            AbandonPending();
            throw cycleError;
#endif
        }
    }

    private void AbandonPending()
    {
        var shouldTerminateSubject = false;
        lock (_gate)
        {
            _pending.Clear();
            shouldTerminateSubject = ReleaseDrainerLocked();
        }
        if (shouldTerminateSubject) TerminateSubject();
    }

    /// <summary>
    /// Releases the active drainer and atomically transfers terminal-stream
    /// ownership to it when disposal arrived during delivery.
    /// Caller must hold <see cref="_gate"/>.
    /// </summary>
    private bool ReleaseDrainerLocked()
    {
        if (_disposeRequested)
        {
            _disposeRequested = false;
            _disposed = true;
            _pending.Clear();
        }
        _drainerThreadId = 0;
        var shouldTerminateSubject = _disposed && !_subjectTerminationClaimed;
        if (shouldTerminateSubject) _subjectTerminationClaimed = true;
        Monitor.PulseAll(_gate);
        return shouldTerminateSubject;
    }

    private void TerminateSubject()
    {
        try
        {
            _subject.OnCompleted();
        }
        finally
        {
            _subject.Dispose();
        }
    }

    /// <summary>Completes and disposes the underlying subject.</summary>
    public void Dispose()
    {
        var caller = Environment.CurrentManagedThreadId;
        var shouldTerminateSubject = false;
        lock (_gate)
        {
            if (_disposed || _disposeRequested) return;
            while (!_disposed && !_disposeRequested)
            {
                var owner = ConflictingOwnerLocked(caller);
                if (owner != 0)
                {
                    if (WaitForOwnerLocked(caller, owner))
                    {
                        _disposeRequested = true;
                        Monitor.PulseAll(_gate);
                        return;
                    }
                    continue;
                }
                if (_borrowedBatchDepth > 0)
                {
                    Monitor.Wait(_gate);
                    continue;
                }
                break;
            }
            if (_disposed || _disposeRequested) return;
            shouldTerminateSubject = FinishDisposeLocked();
            Monitor.PulseAll(_gate);
        }
        if (shouldTerminateSubject) TerminateSubject();
    }

    private int ConflictingOwnerLocked(int caller)
    {
        if (_batchOwnerThreadId != 0 && _batchOwnerThreadId != caller)
            return _batchOwnerThreadId;
        if (_drainerThreadId != 0 && _drainerThreadId != caller)
            return _drainerThreadId;
        return 0;
    }

    private bool WaitForOwnerLocked(int caller, int owner)
    {
        lock (s_waitGraphGate)
        {
            s_waitGraph[caller] = owner;
            var cursor = owner;
            var visited = new HashSet<int>();
            while (s_waitGraph.TryGetValue(cursor, out var next))
            {
                if (next == caller)
                {
                    s_waitGraph.Remove(caller);
                    return true;
                }
                if (!visited.Add(cursor)) break;
                cursor = next;
            }
        }

        try
        {
            Monitor.Wait(_gate);
        }
        finally
        {
            lock (s_waitGraphGate)
                s_waitGraph.Remove(caller);
        }
        return false;
    }

    private bool FinishDisposeLocked()
    {
        _disposeRequested = false;
        _disposed = true;
        _pending.Clear();
        if (_drainerThreadId != 0 || _subjectTerminationClaimed) return false;

        _subjectTerminationClaimed = true;
        return true;
    }
}
